### PR TITLE
[Docs] Extensibility: Permissions Framework

### PR DESCRIPTION
**Pull request by dragon-slayer875**
**Notes for Reviewers**

This PR adds docs for how CASL is used to implement an extensible permissions framework in meshery UI.

**[Signed commits](https://github.com/meshery/meshery/blob/master/CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [x] Yes, I signed my commits.

### CODE REVIEW

**Comment by l5io** [comment]
🚀 Preview for commit 1a7ec0b4e9e3e28d70141b0a0a7056522386be4c at: https://66e98408490b562860f0832d--meshery-docs-preview.netlify.app

File: docs/pages/extensibility/ui.md [hunk 1]
```diff
@@ -39,8 +39,94 @@ Meshery leverages remote providers for identity management. These providers can
...
+### Extensibility: Permissions (CASL)
 
-###  Build-Time UI Extensibility
+Meshery UI uses [CASL.js](https://casl.js.org/v6/en/) to implement it's permissions framework. CASL is an isomorphic authorization JavaScript library which restricts what resources a given client is allowed to access.
+It has been designed to be very customizable and robust.
+It uses keys as it's unit checks to see if a permission to perform or action or not.
+
+#### Implementation
+
+Once a user has logged in, the backend will send a response with the permissions that the user has. Those permissions will be used to check abilities on the frontend.
```
Comment (sudhanshutech) [inline docs/pages/extensibility/ui.md:50]: We should be mentioning that these keys are grouped under keychains and they are assigned to roles then those roles are assigned to users. Then the server checks what roles does user has and they are access to do.  Not directly backend will send a response.

**Review by sudhanshutech — CHANGES_REQUESTED** [review]

File: docs/pages/extensibility/ui.md (hunk 1 above)
Comment (leecalcote) [reply docs/pages/extensibility/ui.md:50]: @sudhanshutech this is great info, however, that functionality is not offered in Meshery directly. Providers can extend Meshery in the way in which you described, if they like, though.

File: docs/pages/extensibility/ui.md [hunk 2]
```diff
@@ -39,8 +39,94 @@ Meshery leverages remote providers for identity management. These providers can
 
 The user avatar behavior, which changes based on the user's status, can be customized by extending the [User Component](https://github.com/meshery/meshery/blob/7de49ef4928f114080f923f2ad261f4433ca91d6/ui/components/User.js#L46).
 
+### Extensibility: Permissions (CASL)
 
-###  Build-Time UI Extensibility
+Meshery UI uses [CASL.js](https://casl.js.org/v6/en/) to implement it's permissions framework. CASL is an isomorphic authorization JavaScript library which restricts what resources a given client is allowed to access.
+It has been designed to be very customizable and robust.
+It uses keys as it's unit checks to see if a permission to perform or action or not.
```
Comment (leecalcote) [inline docs/pages/extensibility/ui.md:46]: "unit checks"?

Comment (leecalcote) [reply docs/pages/extensibility/ui.md:46]: Include one statement about CASL and that's it. This isn't about CASL. This is about Meshery's extensible permission framework, how it works, and how to use it to extend Meshery. How are new keys added? Do they ever change? Where is the current list? Is there an example extension? Why use these?

File: docs/pages/extensibility/ui.md [hunk 3]
```diff
@@ -39,8 +39,94 @@ Meshery leverages remote providers for identity management. These providers can
...
+
+Meshery UI uses `permissions_constants.js` file to keep record of all the keys to check against (fully customizable if self hosting meshery). The keys are accessible using monikers (properties) of the keys object and have been setup this way to make devEx better. Anytime any key
+needs to be added to an action, it needs to be added in this file with an appropriate moniker.
+
+```
+meshery
+└── ui
+    └── utils
+        ├── ...
+        └── permission_constants.js
```
Comment (leecalcote) [inline docs/pages/extensibility/ui.md]: Good of you to include this file tree.

File: docs/pages/extensibility/ui.md [hunk 4]
```diff
@@ -39,8 +39,94 @@ Meshery leverages remote providers for identity management. These providers can
...
+      "function": "View All Users",
+    },
+    {
+      "id": "fa7de118-2d08-4b07-b9d7-3e0baead6d04",
+      "function": "View Profile",
+    },
+]}
+```
+
+Meshery UI uses `permissions_constants.js` file to keep record of all the keys to check against (fully customizable if self hosting meshery). The keys are accessible using monikers (properties) of the keys object and have been setup this way to make devEx better. Anytime any key
```
Comment (leecalcote) [inline docs/pages/extensibility/ui.md]: Always capitalize Meshery.

Comment (leecalcote) [reply docs/pages/extensibility/ui.md]: Don't use "devEx"

Comment (leecalcote) [reply docs/pages/extensibility/ui.md]: Don't describe monikers as properties.

Comment (leecalcote) [reply docs/pages/extensibility/ui.md]: How does the permission framework and its customizability change if Meshery is self-hosted?

**Review by leecalcote — CHANGES_REQUESTED** [review]

File: docs/pages/extensibility/ui.md [hunk 5]
```diff
@@ -39,8 +39,105 @@ Meshery leverages remote providers for identity management. These providers can
...
+
+- It uses keys as it's unit checks to see if a user has permission to perform action or not.
+- These keys are bundled together in keychains. Keychains are collections of similar permissions, which can themselves be grouped together and assigned to roles.
+- Roles map permissions to users. Roles contain any number of keychains, which contain any number of keys (permissions). Meshery uses roles to assign permissions to users.
+
+{% include alert.html type="info" title="Learn more" content="To learn more about how permissions work in Meshery Provider in terms of Roles, keychains and keys, head over to <a href='https://docs.layer5.io/cloud/security/'>Layer5 documentation</a>." %}
+
+#### Implementation
+
+- Once a user has logged in, the backend will send a response with the permissions that the user has, based on the role assigned to them. Those permissions will be used to check abilities on the frontend.
```
Comment (leecalcote) [inline docs/pages/extensibility/ui.md:57]: backend?  Is that Meshery Server or the Provider or a combination?

frontend? You mean Meshery UI? 

When naming these architectural components, include hyperlinks to their concepts pages.

File: docs/pages/extensibility/ui.md [hunk 6]
```diff
@@ -39,8 +39,105 @@ Meshery leverages remote providers for identity management. These providers can
...
+
+{% capture code_content %}Example keys response
```
Comment (leecalcote) [inline docs/pages/extensibility/ui.md:59]: Where is this response stored?

Are cookies used?

File: docs/pages/extensibility/ui.md [hunk 7]
```diff
@@ -39,8 +39,105 @@ Meshery leverages remote providers for identity management. These providers can
...
+RESET_DATABASE: {
+subject: 'Reset database',
+action: '84fc402c-f33e-4a21-a0e3-e14f9e20b125',
+},
+};
+{% endcapture %}
+{% include code.html code=code_content %}
+
+<br>
+- Along with it, we export the `ability.can(action, subject)` provided by CASL as `CAN(action, subject)` for better devEx.
```
Comment (leecalcote) [inline docs/pages/extensibility/ui.md:99]: Don't say "we"

File: docs/pages/extensibility/ui.md [hunk 8]
```diff
@@ -39,8 +39,105 @@ Meshery leverages remote providers for identity management. These providers can
...
+Meshery UI uses [CASL.js](https://casl.js.org/v6/en/) to implement it's permissions framework. CASL is an isomorphic authorization JavaScript library which restricts what resources a given client is allowed to access.
+Meshery's permissions framework has been designed to be very customizable and robust.
+
+- It uses keys as it's unit checks to see if a user has permission to perform action or not.
+- These keys are bundled together in keychains. Keychains are collections of similar permissions, which can themselves be grouped together and assigned to roles.
+- Roles map permissions to users. Roles contain any number of keychains, which contain any number of keys (permissions). Meshery uses roles to assign permissions to users.
+
+{% include alert.html type="info" title="Learn more" content="To learn more about how permissions work in Meshery Provider in terms of Roles, keychains and keys, head over to <a href='https://docs.layer5.io/cloud/security/'>Layer5 documentation</a>." %}
+
+#### Implementation
```
Comment (leecalcote) [inline docs/pages/extensibility/ui.md:55]: Behavior, not Implementation

File: docs/pages/extensibility/ui.md [hunk 9]
```diff
@@ -39,8 +39,105 @@ Meshery leverages remote providers for identity management. These providers can
 
 The user avatar behavior, which changes based on the user's status, can be customized by extending the [User Component](https://github.com/meshery/meshery/blob/7de49ef4928f114080f923f2ad261f4433ca91d6/ui/components/User.js#L46).
 
+### Extensibility: Permissions (CASL)
 
-###  Build-Time UI Extensibility
+#### Design
```
Comment (leecalcote) [inline docs/pages/extensibility/ui.md:44]: Functionalty, not Design

File: docs/pages/extensibility/ui.md [hunk 10]
```diff
@@ -39,8 +39,105 @@ Meshery leverages remote providers for identity management. These providers can
...
+
+Meshery UI uses [CASL.js](https://casl.js.org/v6/en/) to implement it's permissions framework. CASL is an isomorphic authorization JavaScript library which restricts what resources a given client is allowed to access.
+Meshery's permissions framework has been designed to be very customizable and robust.
+
+- It uses keys as it's unit checks to see if a user has permission to perform action or not.
+- These keys are bundled together in keychains. Keychains are collections of similar permissions, which can themselves be grouped together and assigned to roles.
+- Roles map permissions to users. Roles contain any number of keychains, which contain any number of keys (permissions). Meshery uses roles to assign permissions to users.
+
+{% include alert.html type="info" title="Learn more" content="To learn more about how permissions work in Meshery Provider in terms of Roles, keychains and keys, head over to <a href='https://docs.layer5.io/cloud/security/'>Layer5 documentation</a>." %}
```
Comment (leecalcote) [inline docs/pages/extensibility/ui.md:53]: You can include this only as an example extension.

File: docs/pages/extensibility/ui.md [hunk 11]
```diff
@@ -39,8 +39,105 @@ Meshery leverages remote providers for identity management. These providers can
...
 
-###  Build-Time UI Extensibility
+#### Design
+
+Meshery UI uses [CASL.js](https://casl.js.org/v6/en/) to implement it's permissions framework. CASL is an isomorphic authorization JavaScript library which restricts what resources a given client is allowed to access.
+Meshery's permissions framework has been designed to be very customizable and robust.
+
+- It uses keys as it's unit checks to see if a user has permission to perform action or not.
+- These keys are bundled together in keychains. Keychains are collections of similar permissions, which can themselves be grouped together and assigned to roles.
+- Roles map permissions to users. Roles contain any number of keychains, which contain any number of keys (permissions). Meshery uses roles to assign permissions to users.
```
Comment (leecalcote) [inline docs/pages/extensibility/ui.md:51]: > Meshery uses roles to assign permissions to users.

No, it doesn't.

File: docs/pages/extensibility/ui.md [hunk 12]
```diff
@@ -39,8 +39,105 @@ Meshery leverages remote providers for identity management. These providers can
...
+### Extensibility: Permissions (CASL)
 
-###  Build-Time UI Extensibility
+#### Design
+
+Meshery UI uses [CASL.js](https://casl.js.org/v6/en/) to implement it's permissions framework. CASL is an isomorphic authorization JavaScript library which restricts what resources a given client is allowed to access.
+Meshery's permissions framework has been designed to be very customizable and robust.
+
+- It uses keys as it's unit checks to see if a user has permission to perform action or not.
+- These keys are bundled together in keychains. Keychains are collections of similar permissions, which can themselves be grouped together and assigned to roles.
```
Comment (leecalcote) [inline docs/pages/extensibility/ui.md:50]: Meshery does not do this.

File: docs/pages/extensibility/ui.md [hunk 13]
```diff
@@ -39,8 +39,105 @@ Meshery leverages remote providers for identity management. These providers can
...
+
+{% include alert.html type="info" title="Learn more" content="To learn more about how permissions work in Meshery Provider in terms of Roles, keychains and keys, head over to <a href='https://docs.layer5.io/cloud/security/'>Layer5 documentation</a>." %}
+
+#### Implementation
+
+- Once a user has logged in, the backend will send a response with the permissions that the user has, based on the role assigned to them. Those permissions will be used to check abilities on the frontend.
+
+{% capture code_content %}Example keys response
+
+{
```
Comment (leecalcote) [inline docs/pages/extensibility/ui.md:61]: Why is there no reference to https://docs.meshery.io/reference/permissions?

**Review by leecalcote — CHANGES_REQUESTED** [review]

**Comment by l5io** [comment]
🚀 Preview for commit 710bb0e05cae061c30a0a42f6eb91ad99370edab at: https://66f2800fad8dedb8bb673377--meshery-docs-preview.netlify.app

**Comment by l5io** [comment]
🚀 Preview for commit 8bcb3b1c7dcab23f21b39f10382e9c5b487e7131 at: https://66f29146135ee4d7462d3b69--meshery-docs-preview.netlify.app

**Comment by l5io** [comment]
🚀 Preview for commit 710bb0e05cae061c30a0a42f6eb91ad99370edab at: https://66f292e58c84b1e4b2b041e7--meshery-docs-preview.netlify.app

**Comment by l5io** [comment]
🚀 Preview for commit 1f7fc3399ff1de117f22fba0e6ef3a905167fb4c at: https://66f2adef68adfd0c6c3a0b9a--meshery-docs-preview.netlify.app

**Comment by l5io** [comment]
🚀 Preview for commit c6280f65a24d95cfb88a29a37b2a2a30d4c4db96 at: https://66f2aefa21853c0025204a6b--meshery-docs-preview.netlify.app

**Comment by l5io** [comment]
🚀 Preview for commit 34103c268ebb0773c76b87f8148ef6cfa310a567 at: https://66f2b11f236224007b95844f--meshery-docs-preview.netlify.app

**Comment by l5io** [comment]
🚀 Preview for commit 710bb0e05cae061c30a0a42f6eb91ad99370edab at: https://66f2ba6fd1808a0596df9ebe--meshery-docs-preview.netlify.app

**Comment by l5io** [comment]
🚀 Preview for commit 03ea21a63a5897f52b261eb1a0bff10779b9d2bd at: https://66fd1c301c356fb86c63ae70--meshery-docs-preview.netlify.app

**Pull request closed without merging**